tests/test_concurrent_map.cpp

```diff
@@ -0,0 +1,333 @@
+             std::back_inserter(merged));
+  for (auto n : merged) {
+    REQUIRE(skip_list.Contains(n));
+  }
+}
+
+TEST_CASE("Two threads insert the same numbers simultaneously",
+          "[Concurrency]") {
+  auto skip_list = SL<int, int>{};
+
+  auto first_numbers = chunk(10 * kThousand, random(0, kThousand)).get();
+  auto second_numbers =
+      std::vector<int>(std::begin(first_numbers), std::end(first_numbers));
+
+  auto first = std::thread([&]() {
+    for (auto n : first_numbers) {
+      skip_list.Insert(n, n);
+    }
+  });
+
+  auto second = std::thread([&]() {
+    for (auto n : second_numbers) {
+      skip_list.Insert(n, n);
+    }
+  });
+
+  first.join();
+  second.join();
+
+  auto merged = std::vector<int>{};
+  std::merge(std::begin(first_numbers), std::end(first_numbers),
+             std::begin(second_numbers), std::end(second_numbers),
+             std::back_inserter(merged));
+  for (auto n : merged) {
+    REQUIRE(skip_list.Contains(n));
+  }
+}
+
+TEST_CASE("Two threads insert different repeating numbers simultaneously",
+          "[Concurrency]") {
+  auto skip_list = SL<int, int>{};
+
+  auto first_numbers =
+      chunk(10 * kThousand, random(2 * kThousand, 3 * kThousand)).get();
+  auto second_numbers = chunk(10 * kThousand, random(0, kThousand)).get();
+
+  auto first = std::thread([&]() {
+    for (auto n : first_numbers) {
+      skip_list.Insert(n, n);
+    }
+  });
+
+  auto second = std::thread([&]() {
+    for (auto n : second_numbers) {
+      skip_list.Insert(n, n);
+    }
+  });
+
+  first.join();
+  second.join();
+
+  auto merged = std::vector<int>{};
+  std::merge(std::begin(first_numbers), std::end(first_numbers),
+             std::begin(second_numbers), std::end(second_numbers),
+             std::back_inserter(merged));
+  for (auto n : merged) {
+    REQUIRE(skip_list.Contains(n));
+  }
+}
+
+TEST_CASE(
+    "One thread inserts, another is trying to erase non-existent elements",
+    "[Concurrency]") {
+  auto skip_list = SL<int, int>{};
+
+  auto first_numbers = chunk(10 * kThousand, random(0, kThousand)).get();
+  auto second_numbers =
+      chunk(10 * kThousand, random(2 * kThousand, 3 * kThousand)).get();
+
+  auto first = std::thread([&]() {
+    for (auto n : first_numbers) {
+      skip_list.Insert(n, n);
+    }
+  });
+
+  auto second = std::thread([&]() {
+    for (auto n : second_numbers) {
+      skip_list.Erase(n);
+    }
+  });
+
+  first.join();
+  second.join();
+
+  for (auto n : first_numbers) {
+    REQUIRE(skip_list.Contains(n));
+  }
+  for (auto n : second_numbers) {
+    REQUIRE(!skip_list.Contains(n));
+  }
+}
+
+TEST_CASE("One thread inserts and second tries to erase the same element",
+          "[Concurrency]") {
+  auto skip_list = SL<int, int>{};
+
+  int num = 10;
+
+  auto first_numbers = chunk(kThousand, random(num, num)).get();
+  auto second_numbers = chunk(kThousand, random(num, num)).get();
```

Comment:
Maybe add numbers count, thousand may be enough for one thread to finish before the other one even starts.